### ISSUE TITLE
Judgment missing in Paladin damage breakdown 

### ISSUE THREAD

**Issue by veluria** [timhul/ClassicSim#58]
Following #54, I wanted to report this, as I had noticed this issue too. I think it simply boils down to the current rotations (both of them) not using Judgment at all except for putting up the Seal of the Crusader debuff. Hence there's no Judgment of the Command damage in the breakdown, which you should probably be using on CD. 

Also, while we are at it, Exorcism is missing from the rotation if the target is an undead or demon.

**Comment by timhul** [comment]
> Following #54, I wanted to report this, as I had noticed this issue too. I think it simply boils down to the current rotations (both of them) not using Judgment at all except for putting up the Seal of the Crusader debuff. Hence there's no Judgment of Command damage in the breakdown, which one should probably be using on CD.
> 

I would accept a new rotation or a modified rotation that uses Judgement on cooldown. The most important part I think is that there are template rotations that are suitable for the common playstyles, and Judgement on CD will be more mana intensive.

> Also, while we are at it, Exorcism is missing from the rotation if the target is an undead or demon.

Yes, see [this issue for missing Paladin things](https://github.com/timhul/ClassicSim/issues/12). I would like to have added Exorcism back then but it was one of those things that got down-prio'd to finish more general things before the launch.

**Comment by veluria** [comment]
Just to test this, I have added

    <cast_if name="Judgement">
		buff_duration "Seal of Command" greater 0
    </cast_if>

to the rotation, but Judgment of Command still does not appear in the damage breakdown.

**Comment by timhul** [comment]
> Just to test this, I have added
> 
> ```
> <cast_if name="Judgement">
> 	buff duration "Seal of Command" greater 0
> </cast_if>
> ```
> 
> to the rotation, but Judgment of Command still does not appear in the damage breakdown. I guess there is something else missing.

It looks reasonable, and its [Seal::judge_effect](https://github.com/timhul/ClassicSim/blob/3996c945843d8d52b7fa20660420d288e100e385/Class/Paladin/Spells/SealOfCommand.cpp#L55) should deal damage as shown. Please check the Statistics -> Rotation breakdown and see whether the rotational executor actually evaluated as successful (this indicates it did cast Judgement for that specific condition).

**Comment by veluria** [comment]
Strange, it does seem to proc.

![image](https://user-images.githubusercontent.com/1119538/68124023-aab31980-ff0e-11e9-8ad2-5dabcd82d697.png)
![image](https://user-images.githubusercontent.com/1119538/68124063-b999cc00-ff0e-11e9-8d4c-4e59b5967f10.png)
Something else I noticed: When looking at the SotC uptime, it's about 30% for me, which is just to high. JotC renews with auto attacks, so in an ideal fight, you should have almost zero uptime of SotC.

**Comment by timhul** [comment]
The damage is registered as Seal of Command rather than Judgement of Command. This is a limitation in the implementation of Seal::judge_effect, since the ownership of the statistics is the Seal, there is no actual Judgement of Command spell.

As for the uptime it'd be good to observe the combat log but it's currently difficult without recompiling. I have a local development branch that should solve both problems, allowing spells to log any event they want rather than being tied to its origin spell, and it would allow generating a combat log for manual review. The issue is that it is not complete and is a large refactoring so it takes quite a while. For now I'll write a test to make sure JotC is properly renewed and if it isn't I'll fix it.

I'll think about this a bit but you have highlighted two important issues that I think are solvable with a single once-and-for-all solution. It is also possible Judgement will be specifically solved since it is a quite special case.

**Comment by timhul** [comment]
In https://github.com/timhul/ClassicSim/commit/31c73755cab3d3a5804fb2f42deb322cf61018a5 tests were added that verifies auto attacks does indeed refresh the duration of JotC.

**Comment by veluria** [comment]
![image](https://user-images.githubusercontent.com/1119538/68136294-536c7380-ff25-11e9-8731-3afd1d8227c8.png)

… is still sort of confusing to me then.

**Comment by timhul** [comment]
The cause turns out to be the switch to Seal of Command which these tests didn't cover (they only covered AAs with no seal). This significantly changed the uptimes of SoC and SotC (>99% and <1% respectively) and a nice DPS increase.

**Comment by timhul** [comment]
A [new release](https://github.com/timhul/ClassicSim/releases/tag/v0.4-alpha-9) has been created that contains the fix for judge refresh. I suggest you experiment with the Judgement rotation using that one and simply look at the actual damage difference rather than the ability breakdown. The damage sum is recorded properly since it is baked into Seal of Command so it is possible to determine the value of Judgement on CD. Then please make a pull request either with rotation modifications, or as a new separate rotation. I'll continue with the Judgement recorded as a separate entry.

**Comment by veluria** [comment]
Great, thanks for your hard work!

**Comment by timhul** [comment]
> Great, thanks for your hard work!

Thank you! And thank you for your observations and contributions, they are important in such a large project as this.

> One question: Can I somehow check for mana in the rotation? For example, I want to use a certain spell if mana is > 20%, and otherwise just use rank 1.

Percentage resource no, absolute resource yes. See Warrior/Rogue rotations for examples. I'd like to add percentage resources as well but it's always been a case of "nice to have but can work around it".

**Comment by veluria** [comment]
Ok, I'll see as to how much of an optimized rotation I can provide. 

By the way, maybe you remember our talk about slow weapons in #48. Now while this change doesn't change any of the things I said, it certainly makes certain weapons a much more viable alternative.

**Comment by veluria** [comment]
Am I seeing this right that Judgement is causing a GCD as well as is on the GCD in ClassicSim? Because it is neither.

**Comment by timhul** [comment]
> Am I seeing this right that Judgement is causing a GCD as well as is on the GCD in ClassicSim? Because it is neither. That's why you can renew SoC almost instantly after casting Judgement.

You are correct that it does. Does it not inflict a short stance-like GCD of half a second or something or is it truly no GCD at all? Can you really judge and hammer the same instant if you're not on the gcd when you start casting?

**Comment by veluria** [comment]
You cannot judge while casting, but otherwise you can pretty much press two spells simultaneously.

**Comment by timhul** [comment]
> You cannot judge while already casting, but otherwise you can pretty much press two spells simultaneously.

Fixed in e30071eaf384bf29f16f80b83d5b47dd55e58c9e.

**Comment by veluria** [comment]
While changing the rotation and trying to cast judgment on cooldown, I noticed once more that there isn't something quite right yet. The uptime of SotC is still around 20%, which should stay at around 1-2%. Basically, the pre combat action should suffice to apply SotC for the rest of the fight. JotC uptime is also only at 80%, so it does seem to be fading at some point, which should not happen when refreshed on every auto attack. At first I thought JoC was overwriting the debuff somehow, but looking at the code, this doesn't seem to be the case. Any ideas?
